hadoop-hdfs-project/hadoop-hdfs-rbf/src/test/java/org/apache/hadoop/hdfs/rbfbalance/TestRouterDistCpProcedure.java

```diff
@@ -0,0 +1,120 @@
+
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.util.Collections;
+
+import static org.apache.hadoop.hdfs.server.federation.FederationTestUtils.createNamenodeReport;
+import static org.apache.hadoop.test.LambdaTestUtils.intercept;
+import static org.junit.Assert.assertTrue;
+
+
+public class TestRouterDistCpProcedure extends TestDistCpProcedure {
+    private static StateStoreDFSCluster cluster;
+    private static MiniRouterDFSCluster.RouterContext routerContext;
+    private static Configuration routerConf;
+    private static StateStoreService stateStore;
+
+    @BeforeClass
+    public static void globalSetUp() throws Exception {
+        cluster = new StateStoreDFSCluster(false, 1);
+        // Build and start a router with State Store + admin + RPC
+        Configuration conf = new RouterConfigBuilder()
+                .stateStore()
+                .admin()
+                .rpc()
+                .build();
+        cluster.addRouterOverrides(conf);
+        cluster.startRouters();
+        routerContext = cluster.getRandomRouter();
+        Router router = routerContext.getRouter();
+        stateStore = router.getStateStore();
+
+        // Add one name services for testing
+        ActiveNamenodeResolver membership = router.getNamenodeResolver();
+        membership.registerNamenode(createNamenodeReport("ns0", "nn1",
+                HAServiceProtocol.HAServiceState.ACTIVE));
+        stateStore.refreshCaches(true);
+
+        routerConf = new Configuration();
+        InetSocketAddress routerSocket = router.getAdminServerAddress();
+        routerConf.setSocketAddr(RBFConfigKeys.DFS_ROUTER_ADMIN_ADDRESS_KEY,
+                routerSocket);
+    }
+
+    @Override
+    public void testDisableWrite() throws Exception {
+        // Firstly add mount entry: /test-write->{ns0,/test-write}.
+        String mount = "/test-write";
+        MountTable newEntry = MountTable
+                .newInstance(mount, Collections.singletonMap("ns0", mount),
+                        Time.now(), Time.now());
+        MountTableManager mountTable =
+                routerContext.getAdminClient().getMountTableManager();
+        AddMountTableEntryRequest addRequest =
+                AddMountTableEntryRequest.newInstance(newEntry);
+        AddMountTableEntryResponse addResponse =
+                mountTable.addMountTableEntry(addRequest);
+        assertTrue(addResponse.getStatus());
+        stateStore.loadCache(MountTableStoreImpl.class, true); // load cache.
+
+        // Construct client.
+        URI address = routerContext.getFileSystemURI();
+        DFSClient routerClient = new DFSClient(address, routerConf);
+
+        FedBalanceContext context = new FedBalanceContext
+                .Builder(null, null, mount, routerConf).build();
+        RouterDistCpProcedure dcProcedure = new RouterDistCpProcedure();
+        executeProcedure(dcProcedure, Stage.FINAL_DISTCP,
+                () -> dcProcedure.disableWrite(context));
```

Comment:
I think we can set the stage to Stage.DISABLE_WRITE and then call ()->dcProcedure.execute() instead of calling dcProcedure.disableWrite(context).  This is to test whether DistCpProcedure handles Stage.DISABLE_WRITE as expected.

Reply:
Thanks for your advice. The TestRouterDistCpProcedure extends TestDistCpProcedure and it includes super class test methods to test the whole process. Maybe, it's better to focus on the disableWrite method result itself in this unit test.